File: zena_mode/resource_manager.py
# -*- coding: utf-8 -*-
"""
resource_manager.py - Dynamic Resource Management for ZenAI V2
Monitors system RAM and determines model loading strategies to prevent crashes.
"""

import logging
from typing import Dict, Literal, Optional
import threading

# from typing import List, Optional
from config_system import config
import asyncio

# Configure logging
logger = logging.getLogger("ResourceManager")

# Try to import psutil, handled gracefully if missing
try:
    import psutil

    PSUTIL_AVAILABLE = True
except ImportError:
    PSUTIL_AVAILABLE = False
    logger.warning("[ResourceManager] psutil not found. Dynamic RAM detection disabled (defaulting to Safe Mode).")


class ResourceManager:
    """
    Monitors system resources and delegates loading strategies.

    Strategies:
    - SERIAL: Unload current -> Load new (Safe, Slow). Used when RAM < 8GB.
    - SWAP: Keep Orchestrator, swap Experts. Used for 8-16GB RAM.
    - PARALLEL: Keep Orchestrator + Expert loaded. Used for > 16GB RAM.
    """

    def __init__(self):
        """Initialize instance."""
        self.strategy: Literal["SERIAL", "SWAP", "PARALLEL"] = "SERIAL"
        self.total_ram_gb = 0
        self.available_ram_gb = 0
        self._refresh_stats()
        self._determine_strategy()
        # Optional worker thread tracking (for background tasks spawned by orchestrator)
        self._worker_threads: List[threading.Thread] = []
        self._threads_lock = threading.Lock()

    def add_worker_thread(self, target, args=(), daemon=True, max_workers: Optional[int] = None):
        """Start and register a background thread, optionally enforcing a max worker count.

        Returns the Thread object.
        """
        # Determine effective max_workers
        max_workers if max_workers is not None else getattr(config, "GENERIC_MAX_WORKERS", 4)

        t = threading.Thread(target=target, args=args, daemon=daemon)
        with self._threads_lock:
            self._worker_threads.append(t)
            # If max_workers is set, prune finished threads and optionally refuse new threads
            if max_workers is not None:
                # cleanup first
                self.cleanup_finished_threads()
                if len(self._worker_threads) >= max_workers:
                    # Do not start new worker to avoid resource exhaustion
                    raise RuntimeError("Max worker threads reached")
        t.start()
        return t

    def cleanup_finished_threads(self):
        """Remove finished worker threads from internal registry."""
        with self._threads_lock:
            self._worker_threads = [t for t in self._worker_threads if t.is_alive()]

    def get_worker_count(self) -> int:
        with self._threads_lock:
            self.cleanup_finished_threads()
            return len(self._worker_threads)

    def run_in_thread_future(self, func, *args, daemon=True, max_workers: Optional[int] = None):
        """Run `func(*args)` in a background thread, return an asyncio.Future tied to the current loop.

        The thread is tracked in the internal registry so cleanup can prune finished threads.
        Raises RuntimeError if max_workers would be exceeded.
        """
        try:
            loop = asyncio.get_running_loop()
        except RuntimeError:
            # Not called from an event loop
            raise RuntimeError("run_in_thread_future must be called from an async context")

        fut = loop.create_future()

        def _target():
            try:
                res = func(*args)
            except Exception as e:
                loop.call_soon_threadsafe(fut.set_exception, e)
                return
            loop.call_soon_threadsafe(fut.set_result, res)

        # Determine effective max_workers
        effective_max = max_workers if max_workers is not None else getattr(config, "GENERIC_MAX_WORKERS", 4)

        t = threading.Thread(target=_target, daemon=daemon)
        with self._threads_lock:
            self._worker_threads.append(t)
            if effective_max is not None:
                self.cleanup_finished_threads()
                if len(self._worker_threads) > effective_max:
                    # remove the appended thread to keep state consistent
                    self._worker_threads.pop()
                    raise RuntimeError("Max worker threads reached")

        t.start()
        return fut

    def _refresh_stats(self):
        """Update current memory stats."""
        if PSUTIL_AVAILABLE:
            mem = psutil.virtual_memory()
            self.total_ram_gb = mem.total / (1024**3)
            self.available_ram_gb = mem.available / (1024**3)
        else:
            # Fallback for no psutil
            self.total_ram_gb = 8.0  # Assume baseline
            self.available_ram_gb = 4.0

    def _determine_strategy(self):
        """Calculate best strategy based on Total RAM."""
        # Note: We use Total RAM for strategy definition, but check Available RAM for immediate actions.

        if self.total_ram_gb < 8.5:  # < 8GB (giving some margin)
            self.strategy = "SERIAL"
        elif self.total_ram_gb < 16.5:  # 8GB - 16GB
            self.strategy = "SWAP"
        else:  # > 16GB
            self.strategy = "PARALLEL"

        logger.info(
            f"[ResourceManager] Strategy: {self.strategy} (Total: {self.total_ram_gb:.1f}GB, Avail: {self.available_ram_gb:.1f}GB)"
        )

    def can_load_model(self, model_size_gb: float) -> bool:
        """Check if we have enough RAM to load a model of size_gb."""
        self._refresh_stats()
        # Leave 2GB buffer for OS/System
        safe_margin = 2.0
        return self.available_ram_gb > (model_size_gb + safe_margin)

    def get_status(self) -> Dict:
        """Return current status for UI."""
        self._refresh_stats()
        return {
            "strategy": self.strategy,
            "total_ram": f"{self.total_ram_gb:.1f} GB",
            "available_ram": f"{self.available_ram_gb:.1f} GB",
            "psutil": PSUTIL_AVAILABLE,
        }


# Global singleton
resource_manager = ResourceManager()

if __name__ == "__main__":
    # Test run
    # [X-Ray auto-fix] print(f"Strategy: {resource_manager.strategy}")
    pass
    # [X-Ray auto-fix] print(f"Can load 4GB model? {resource_manager.can_load_model(4.0)}")    pass

File: rust_output/src/resource_manager.rs
/// resource_manager::py - Dynamic Resource Management for ZenAI V2
/// Monitors system RAM and determines model loading strategies to prevent crashes.

use anyhow::{Result, Context};
use crate::config_system::{config};
use std::collections::HashMap;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

pub static RESOURCE_MANAGER: std::sync::LazyLock<ResourceManager> = std::sync::LazyLock::new(|| Default::default());

/// Monitors system resources and delegates loading strategies.
/// 
/// Strategies:
/// - SERIAL: Unload current -> Load new (Safe, Slow). Used when RAM < 8GB.
/// - SWAP: Keep Orchestrator, swap Experts. Used for 8-16GB RAM.
/// - PARALLEL: Keep Orchestrator + Expert loaded. Used for > 16GB RAM.
#[derive(Debug, Clone)]
pub struct ResourceManager {
    pub strategy: Literal<serde_json::Value>,
    pub total_ram_gb: i64,
    pub available_ram_gb: i64,
    pub _worker_threads: Vec<threading::Thread>,
    pub _threads_lock: std::sync::Mutex<()>,
}

impl ResourceManager {
    /// Initialize instance.
    pub fn new() -> Self {
        Self {
            strategy: Default::default(),
            total_ram_gb: 0,
            available_ram_gb: 0,
            _worker_threads: Vec::new(),
            _threads_lock: std::sync::Mutex::new(()),
        }
    }
    /// Start and register a background thread, optionally enforcing a max worker count.
    /// 
    /// Returns the Thread object.
    pub fn add_worker_thread(&mut self, target: String, args: String, daemon: String, max_workers: Option<i64>) -> Result<()> {
        // Start and register a background thread, optionally enforcing a max worker count.
        // 
        // Returns the Thread object.
        if max_workers.is_some() { max_workers } else { /* getattr */ 4 };
        let mut t = std::thread::spawn(|| {});
        let _ctx = self._threads_lock;
        {
            self._worker_threads.push(t);
            if max_workers.is_some() {
                self.cleanup_finished_threads();
                if self._worker_threads.len() >= max_workers {
                    return Err(anyhow::anyhow!("RuntimeError('Max worker threads reached')"));
                }
            }
        }
        t.start();
        Ok(t)
    }
    /// Remove finished worker threads from internal registry.
    pub fn cleanup_finished_threads(&mut self) -> () {
        // Remove finished worker threads from internal registry.
        let _ctx = self._threads_lock;
        {
            self._worker_threads = self._worker_threads.iter().filter(|t| t.is_alive()).map(|t| t).collect::<Vec<_>>();
        }
    }
    pub fn get_worker_count(&self) -> i64 {
        let _ctx = self._threads_lock;
        {
            self.cleanup_finished_threads();
            self._worker_threads.len()
        }
    }
    /// Run `func(*args)` in a background thread, return an asyncio.Future tied to the current loop.
    /// 
    /// The thread is tracked in the internal registry so cleanup can prune finished threads.
    /// Raises RuntimeError if max_workers would be exceeded.
    pub fn run_in_thread_future(&mut self, func: String, args: Vec<Box<dyn std::any::Any>>) -> Result<()> {
        // Run `func(*args)` in a background thread, return an asyncio.Future tied to the current loop.
        // 
        // The thread is tracked in the internal registry so cleanup can prune finished threads.
        // Raises RuntimeError if max_workers would be exceeded.
        // try:
        {
            let mut r#loop = asyncio.get_running_loop();
        }
        // except RuntimeError as _e:
        let mut fut = r#loop.create_future();
        let _target = || {
            // try:
            {
                let mut res = func(/* *args */);
            }
            // except Exception as e:
            r#loop.call_soon_threadsafe(fut.set_result, res);
        };
        let mut effective_max = if max_workers.is_some() { max_workers } else { /* getattr */ 4 };
        let mut t = std::thread::spawn(|| {});
        let _ctx = self._threads_lock;
        {
            self._worker_threads.push(t);
            if effective_max.is_some() {
                self.cleanup_finished_threads();
                if self._worker_threads.len() > effective_max {
                    self._worker_threads.pop().unwrap();
                    return Err(anyhow::anyhow!("RuntimeError('Max worker threads reached')"));
                }
            }
        }
        t.start();
        Ok(fut)
    }
    /// Update current memory stats.
    pub fn _refresh_stats(&mut self) -> () {
        // Update current memory stats.
        if PSUTIL_AVAILABLE {
            let mut mem = psutil.virtual_memory();
            self.total_ram_gb = (mem.total / (1024).pow(3 as u32));
            self.available_ram_gb = (mem.available / (1024).pow(3 as u32));
        } else {
            self.total_ram_gb = 8.0_f64;
            self.available_ram_gb = 4.0_f64;
        }
    }
    /// Calculate best strategy based on Total RAM.
    pub fn _determine_strategy(&mut self) -> () {
        // Calculate best strategy based on Total RAM.
        if self.total_ram_gb < 8.5_f64 {
            self.strategy = "SERIAL".to_string();
        } else if self.total_ram_gb < 16.5_f64 {
            self.strategy = "SWAP".to_string();
        } else {
            self.strategy = "PARALLEL".to_string();
        }
        logger.info(format!("[ResourceManager] Strategy: {} (Total: {:.1}GB, Avail: {:.1}GB)", self.strategy, self.total_ram_gb, self.available_ram_gb));
    }
    /// Check if we have enough RAM to load a model of size_gb.
    pub fn can_load_model(&mut self, model_size_gb: f64) -> bool {
        // Check if we have enough RAM to load a model of size_gb.
        self._refresh_stats();
        let mut safe_margin = 2.0_f64;
        self.available_ram_gb > (model_size_gb + safe_margin)
    }
    /// Return current status for UI.
    pub fn get_status(&self) -> HashMap {
        // Return current status for UI.
        self._refresh_stats();
        HashMap::from([("strategy".to_string(), self.strategy), ("total_ram".to_string(), format!("{:.1} GB", self.total_ram_gb)), ("available_ram".to_string(), format!("{:.1} GB", self.available_ram_gb)), ("psutil".to_string(), PSUTIL_AVAILABLE)])
    }
}
